### PR TITLE
leave lobby handler

### LINKED ISSUE #106: Leave Lobby button
Players need to be able to quit lobbies. The 'Leave' button should tell the server that the player left the game.

### PR DESCRIPTION
**Pull request by MondayHopscotch**
Fixes #106

### CODE REVIEW

File: server/routing/lobby.go [hunk 1]
```diff
@@ -137,15 +164,18 @@ func (lr *LobbyRoutes) lobbyStartGameHandler(w http.ResponseWriter, r *http.Requ
...
 	}
 
-	// tell all pubsubbers
+	w.Write([]byte("The game has started."))
+
+	PublishMessage(msg, lobby)
+	return
+}
+
+func PublishMessage(msg pogo.Typer, lobby services.Lobby) {
```
Comment (Kenoshen) [inline server/routing/lobby.go]: I like this encapsulation a lot.  Thank you for doing that.  I knew we needed to, but I wasn't exactly sure how to structure it.

File: server/routing/lobby_test.go [hunk 2]
```diff
@@ -196,3 +196,60 @@ func TestBadLobbyRequest(t *testing.T) {
...
+	}
+
+	_, err = put(port, leaveEndpoint, leaveMsg)
+	if !assert.Nil(t, err) {
+		t.Fatal(err)
+	}
+
+	lobbies = svcs.Lobby.GetLobbiesSnapshot()
+	if !assert.Len(t, lobbies[lobbyID].Players, 0) {
+		t.Fatal("expected 0 players in game lobby")
```
Comment (Kenoshen) [inline server/routing/lobby_test.go]: When there are 0 players in a lobby, should the lobby kill itself?  I can't remember, are there separate steps to creating and then joining a lobby?  Or is that one combined step?  If it is separate, then I guess having 0 players in the lobby is how it starts when it is first created.  But if it is joined, then the default lobby always has at least 1 player in it.  So if we allow players to leave lobbies now, we should probably handle an empty lobby.

File: server/services/lobby_service.go [hunk 3]
```diff
@@ -84,6 +85,33 @@ func (l *LobbyServiceImpl) JoinLobby(msg pogo.LobbyJoinMsg) (Lobby, error) {
...
+			playerNum = i
+			break
+		}
+	}
+
+	if playerNum == -1 {
+		return Lobby{}, fmt.Errorf("lobby '%v' does not have player '%v'", lobbyID, msg.PlayerID)
+	}
+
+	delete(lobby.ActiveConnections, msg.PlayerID)
```
Comment (Kenoshen) [inline server/services/lobby_service.go]: Does this actually kill the websocket connection?

File: server/routing/lobby_test.go (hunk 2 above)
Comment (MondayHopscotch) [reply server/routing/lobby_test.go]: Yeah, it's two steps right now to create and join. But it still probably makes sense to close the lobby after the last player LEAVES it.

File: server/services/lobby_service.go (hunk 3 above)
Comment (MondayHopscotch) [reply server/services/lobby_service.go]: hmm... no, no it doesn't. I'll add an actual call to close the socket.

**Pull request merged**